### ISSUE TITLE
Blacklisting IP's from file

### ISSUE THREAD

**Issue by arakis** [jcmoraisjr/haproxy-ingress#509]
Hello,
how can i use files like:
```
        # deny access from specific countries
        acl src_blacklist src -f /haproxy-config/blocked-ip-ranges.csv
        tcp-request connection reject if src_blacklist
```
My guess is, this is possible via config maps. How to reference them with the ` -f` argument?

**Comment by jcmoraisjr** [comment]
Hi, in theory, if you deploy haproxy-ingress as a k8s object, you can:

1. create a configmap with this blacklist, one IP or CIDR on its own line
2. mount this configmap as a volume in the haproxy-ingress' pod template
3. reference the mounted dir/file using `-f /dir/file`
4. such -f configuration would be in the `config-backend` annotation or global configmap option

Note also that changing the blacklist wouldn't update haproxy, you'd need to wait the file content to be updated (this could happen in 10 seconds to 1 minute) and then trigger a reload eg updating max-connection with +/-1

`whitelist-source-range` in the global configmap is also an option but will create the same acl on every backend of your cfg.

**Comment by arakis** [comment]
I successfully mounted a file /blocked-ip-ranges.csv (it's x.x.x.x/y format, one per line) into the pods of haproxy-ingress in kubernetes, checked via `kubectl exec -it --namespace ingress-controller haproxy-ingress-xxxxx -- cat /blocked-ip-ranges.csv`

Now, on the deployment of a specific service (consumer of haproxy-ingress), i added a "ingress.kubernetes.io/config-frontend"-section:

```
  annotations:
    kubernetes.io/ingress.class: "haproxy"

    ingress.kubernetes.io/config-frontend: |
      # deny access from specific countries
      acl src_blacklist src -f /blocked-ip-ranges.csv
      tcp-request connection reject if src_blacklist
```
It seems nothing is blocked. I guess the reason is the google loadbalancer, maybe the src is that IP? How to archive that the client's IP is blocked?

Btw, a `blacklist-source-range` option would be very usefull.

**Comment by arakis** [comment]
i also tried `req.hdr_ip(X-Forwarded-For, 1)` instead of `src`. I know, that the my production pods are receiving `x-forwarded-for=<client-ip>, <another-ip>`, so even the `-1` should be correct.

I'm using docker image v.0.7.5.

**Comment by arakis** [comment]
I was able to get it working:

mount config map as described above. You can do this with a small script like:
```
kubectl delete configmap blocked-ip-ranges --namespace ingress-controller
kubectl create configmap blocked-ip-ranges --from-file=blocked-ip-ranges.csv --namespace ingress-controller
```

Than in your Ingress, use this:

```
ingress.kubernetes.io/config-backend: |
  # Blacklist
  acl src_blacklist req.hdr_ip(X-Forwarded-For, -1) -f /blocked-ip-ranges.csv
  http-request deny deny_status 500 if src_blacklist
```
The ` -1` is requied because `X-Forwarded-For` contains two IPs: "`<client-ip> <google-load-balancer-ip>`.

My only question would be, how to update the file? My current workflow is, updating the configmap and then force re-deploy of the haproxy-ingress deployment, but change a little field like an unused Environment-Variable. I'm not sure if there's an more elegant way.

**Comment by jcmoraisjr** [comment]
Hi, thanks for the feedback and sorry the long delay. Vacation here.

You can also patch the configmap, k8s will update the file on disk. You need however to check when the file is updated before reload haproxy because the update is async and might take between 10 and 30 seconds to be updated.

You can force a soft reload of haproxy changing any configuration that need a haproxy reload to be applied. You can change eg max-connections +/- 1.

Just to let you know I'll add this feature to haproxy ingress, perhaps in the v0.10 which is the current master.

**Comment by thyn** [comment]
Same problems. I've created file and everything is fine besides https ssl passthrough. The problem is that I don't know how to insert custom code to https tcp frontend. All passthrough sites are not filtered. It works only in regular https

`{
	"config-frontend": "acl network_allowed src -f /etc/haproxy/lists/allowedips.lst
		http-request deny if !network_allowed
		"
}
`

**Comment by jcmoraisjr** [comment]
Hi @thyn please file a new issue with the proposal so this won't be forgotten. This fr relates to a blacklist source range

**Comment by jcmoraisjr** [comment]
@thyn nevermind, sorry about the noise, I just found #587 here.

**Comment by jcmoraisjr** [comment]
Implemented in v0.12-beta.1. Closing.
